Readme.md
# PCG Random Number Generation, Java Edition


<a href="https://travis-ci.org/KilianB/pcg-java"><img src="https://travis-ci.org/KilianB/pcg-java.svg?branch=master"/></a>
<a href="https://github.com/KilianB/pcg-java/blob/master/LICENSE"><img src="https://img.shields.io/badge/license-MIT%20License-blue.svg"/></a>
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/72dd6140019e4658bc2bc21830271175)](https://app.codacy.com/app/KilianB/pcg-java?utm_source=github.com&utm_medium=referral&utm_content=KilianB/pcg-java&utm_campaign=Badge_Grade_Dashboard)
[![Codecov](https://img.shields.io/codecov/c/github/codecov/example-python.svg)](https://codecov.io/gh/KilianB/pcg-java)



> PCG is a family of simple fast space-efficient statistically good algorithms for random number generation. Unlike many general-purpose RNGs, they are also hard to predict.


This implementation is based on Melissa E. O'Neill <a href="https://github.com/imneme/pcg-c">C code</a> and <a href="http://www.pcg-random.org/paper.html">paper</a> <a href="http://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf">PCG: A Family of Simple Fast
Space-Efficient Statistically Good Algorithms for Random Number Generation</a>. 

You are highly advised to check out her website for further information: <a href="http://www.pcg-random.org">www.pcg-random.org</a>
Java implementation featuring a 64 bit state RNG with 32 bit output.

<p align="center">
	<img src ="https://user-images.githubusercontent.com/9025925/45636036-e2f75e80-baa6-11e8-9db8-f43c2c709943.png">
</p>
Table taken from <a href="http://www.pcg-random.org/">http://www.pcg-random.org/</a>.


## Features

<ul>
	<li>High quality number generator</li>
	<li>Thread safe</li>
	<li>Quick advance and rewind allowing to reset the generator to a previous state or to skip the next k numbers</li>
	<li>Calculate the distance between different generator instances</li>
	<li>Extends the random interface to conveniently uses methods like <code>Collection.sort</code></li>
	<li>Allows to split instances like <code>SplittableRandom</code></li>
</ul>

Note: The PCG family is <b>not</b> cryptographically secure! 

## Usage 

1. Either download the <a href="build">pre build jar files</a> and add them to your build path manually
2. Use Maven, gradle or ivy via maven central 


````
<!-- Java 10 Version -->
<dependency>
  <groupId>dev.brachtendorf</groupId>
  <artifactId>pcg-java-10</artifactId>
  <version>1.0.1</version>
</dependency>

<!-- Java 8 Version -->
<dependency>
  <groupId>dev.brachtendorf</groupId>
  <artifactId>pcg-java-8</artifactId>
  <version>1.0.1</version>
</dependency>

````



Several different random number generators are bundled in this package. Pick one which suits your needs

````Java

// Uniquely seed
PcgRR rng = new PcgRR();

int randomInt = rng.nextInt();
double nextGaus = rng.nextGaus();

````

<details>
  
  <summary>Exand additional features and examples.</summary>
  
````Java
/* Next to randomly setting the internal state (seeding) the pcg family also allows to select one of 
many streams which all follow a different random number sequence ensuring that internal stages 
never  converge between instance. */
long seed = 0;
long streamNumber = 0;
PcgRR rng = new PcgRR(seed,streamNumber);

/* This istance will produce the same output when generating a number but they don't share a
reference*/
PcgRR rngClone = rng.split();

rng.nextInt() == rngClone.nextInt() // TRUE

/* This instance does not share a state and has a different seed and stream number!*/
RngRR distinctRng = rng.splitDistinct();

//Distance
rngClone = rng.split();

for(int i = 0; i < 100; i++;){
    rng.nextInt();
}
rngClone.distance(rng)  // 100

//Rewind or fast skip x numbers
int randomInt = rng.nextInt()
rng.advance(-1) // rewind 1 step 
randomInt == rng.Next() // TRUE
````

</details>
<br>


### Choose the correct generator 
<table>
	<tr>
		<td>Generator</td>
		<td>Description</td>
		<td>Thread safety</td>
		<td>Notes</td>
	</tr>
	<tr>	
		<td>PcgRR</td>
		<td>Balance speed with statistical performance and reasonable security.
		See chapter 6.3.1 in the paper. </td>
		<td>Yes synchronization*</td>
		<td>Recommended for multi threaded use</td>
	</tr>
	<tr>
		<td>PcgRS</td>
		<td>Instead of rotating the bits perform an implied shift.
			See chapter 6.3.2 in the paper.
			Slightly worse statistical performance**
		</td>
		<td>Yes synchronization*</td>
		<td>This generator is supposed to be faster due to performing a shift instead of rotating bits. But in benchmarks it performed slightly worse than the RR version. No real reason to pick this one over the RR at the moment.</td>
	</tr>
	<tr>
		<td>PcgRSFast</td>
		<td>
<ul>
<li>Identical output to the PcgRS but drops synchronization</li>
<li>Manually inlined all method calls and use bit masking instead of casting where ever applicable</li>
</ul>
		</td>		
		<td>No!</td>
		<td>The go to instance when thread safety is not desired</td>
	</tr>
	<tr>
		<td>PcgRSUFast</td>
		<td>
			<ul>
				<li>highest performance overall and identical code to the PcgRSFast but additionally
			uses static methods instead of local once</li>
				<li>Does not extend java.util.Random</li>
			</ul>
		</td>
		<td>No!</td>
		<td>This is a more lets see
			how quick we can get approach instead of a reasonable class. Usually the PcgRSFast is preferable.</td>
	</tr>
</table>

* 3 separate implementations with different synchronization approaches were implemented and benchmarked with synchronization outperforming
locks and cas instructions by far.
** when speaking about less statistical performance we are still way beyond everything generated by java random or mt!

## Benchmark
Statistical and performance benchmarks were performed using pract rand and Oracle's jmh harness respectively 

### Statistical Quality (performed on nextInt only)

PractRand Test Suite

````
rng=PcgRSUFast stdin32, seed=Random
length= 64 gigabytes (2^36 bytes), time= 2351 seconds
  Test Name                         Raw       Processed     Evaluation
  BCFN(2+0,13-0,T)                  R=  +1.0  p = 0.335     normal           
  BCFN(2+1,13-0,T)                  R=  -2.7  p = 0.865     normal           
  BCFN(2+2,13-0,T)                  R=  +0.7  p = 0.386     normal           
````

<br>
<details>

<summary>Click here to expand the full report</summary>

````
rng=PcgRSUFast stdin32, seed=Random
length= 64 gigabytes (2^36 bytes), time= 2351 seconds
  Test Name                         Raw       Processed     Evaluation
  BCFN(2+0,13-0,T)                  R=  +1.0  p = 0.335     normal           
  BCFN(2+1,13-0,T)                  R=  -2.7  p = 0.865     normal           
  BCFN(2+2,13-0,T)                  R=  +0.7  p = 0.386     normal           
  BCFN(2+3,13-0,T)                  R=  +4.2  p = 0.044     normal           
  BCFN(2+4,13-0,T)                  R=  +1.2  p = 0.315     normal           
  BCFN(2+5,13-0,T)                  R=  -0.8  p = 0.627     normal           
  BCFN(2+6,13-0,T)                  R=  +1.8  p = 0.233     normal           
  BCFN(2+7,13-0,T)                  R=  -2.3  p = 0.824     normal           
  BCFN(2+8,13-1,T)                  R=  -1.9  p = 0.781     normal           
  BCFN(2+9,13-1,T)                  R=  -4.7  p = 0.976     normal           
  BCFN(2+10,13-2,T)                 R=  +3.5  p = 0.078     normal           
  BCFN(2+11,13-3,T)                 R=  +2.7  p = 0.136     normal           
  BCFN(2+12,13-3,T)                 R=  -3.7  p = 0.944     normal           
  BCFN(2+13,13-4,T)                 R=  -0.3  p = 0.532     normal           
  BCFN(2+14,13-5,T)                 R=  -0.4  p = 0.543     normal           
  BCFN(2+15,13-5,T)                 R=  +1.6  p = 0.236     normal           
  BCFN(2+16,13-6,T)                 R=  -0.2  p = 0.494     normal           
  BCFN(2+17,13-6,T)                 R=  +1.5  p = 0.243     normal           
  BCFN(2+18,13-7,T)                 R=  -2.0  p = 0.793     normal           
  BCFN(2+19,13-8,T)                 R=  -3.2  p = 0.960     normal           
  BCFN(2+20,13-8,T)                 R=  -3.1  p = 0.956     normal           
  BCFN(2+21,13-9,T)                 R=  -2.3  p = 0.891     normal           
  BCFN(2+22,13-9,T)                 R=  -2.1  p = 0.860     normal           
  DC6-9x1Bytes-1                    R=  +1.9  p = 0.191     normal           
  Gap-16:A                          R=  +0.5  p = 0.435     normal           
  Gap-16:B                          R=  +0.9  p = 0.270     normal           
  FPF-14+6/16:(0,14-0)              R=  +3.2  p = 0.014     normal           
  FPF-14+6/16:(1,14-0)              R=  -1.2  p = 0.806     normal           
  FPF-14+6/16:(2,14-0)              R=  -1.5  p = 0.861     normal           
  FPF-14+6/16:(3,14-0)              R=  -1.0  p = 0.751     normal           
  FPF-14+6/16:(4,14-0)              R=  +0.5  p = 0.356     normal           
  FPF-14+6/16:(5,14-0)              R=  +1.4  p = 0.158     normal           
  FPF-14+6/16:(6,14-0)              R=  -1.3  p = 0.825     normal           
  FPF-14+6/16:(7,14-0)              R=  -0.5  p = 0.638     normal           
  FPF-14+6/16:(8,14-0)              R=  -0.1  p = 0.538     normal           
  FPF-14+6/16:(9,14-0)              R=  -0.4  p = 0.609     normal           
  FPF-14+6/16:(10,14-0)             R=  +0.6  p = 0.331     normal           
  FPF-14+6/16:(11,14-0)             R=  +0.2  p = 0.433     normal           
  FPF-14+6/16:(12,14-1)             R=  +0.1  p = 0.467     normal           
  FPF-14+6/16:(13,14-2)             R=  +0.3  p = 0.421     normal           
  FPF-14+6/16:(14,14-2)             R=  +0.3  p = 0.427     normal           
  FPF-14+6/16:(15,14-3)             R=  +1.2  p = 0.198     normal           
  FPF-14+6/16:(16,14-4)             R=  +0.6  p = 0.336     normal           
  FPF-14+6/16:(17,14-5)             R=  -1.8  p = 0.904     normal           
  FPF-14+6/16:(18,14-5)             R=  -2.6  p = 0.973     normal           
  FPF-14+6/16:(19,14-6)             R=  +0.4  p = 0.380     normal           
  FPF-14+6/16:(20,14-7)             R=  -0.8  p = 0.713     normal           
  FPF-14+6/16:(21,14-8)             R=  +3.1  p = 0.023     normal           
  FPF-14+6/16:(22,14-8)             R=  +0.8  p = 0.281     normal           
  FPF-14+6/16:(23,14-9)             R=  +0.0  p = 0.464     normal           
  FPF-14+6/16:(24,14-10)            R=  -1.6  p = 0.888     normal           
  FPF-14+6/16:(25,14-11)            R=  -0.9  p = 0.694     normal           
  FPF-14+6/16:(26,14-11)            R=  -0.4  p = 0.539     normal           
  FPF-14+6/16:all                   R=  +0.1  p = 0.478     normal           
  FPF-14+6/16:cross                 R=  +1.4  p = 0.091     normal           
  BRank(12):128(16)                 R=  -0.1  p~= 0.520     normal           
  BRank(12):256(8)                  R=  -0.2  p~= 0.490     normal           
  BRank(12):384(2)                  R=  +1.6  p~= 0.168     normal           
  BRank(12):512(8)                  R=  +2.5  p~= 0.020     normal           
  BRank(12):768(2)                  R=  -0.2  p~= 0.554     normal           
  BRank(12):1K(4)                   R=  +0.4  p~= 0.340     normal           
  BRank(12):1536(1)                 R=  -0.7  p~= 0.689     normal           
  BRank(12):2K(4)                   R=  +1.7  p~= 0.060     normal           
  BRank(12):3K(1)                   R=  +1.8  p~= 0.146     normal           
  BRank(12):4K(2)                   R=  -0.2  p~= 0.554     normal           
  BRank(12):6K(1)                   R=  -0.7  p~= 0.689     normal           
  BRank(12):8K(2)                   R=  +0.8  p~= 0.293     normal           
  BRank(12):12K(1)                  R=  -0.7  p~= 0.689     normal           
  mod3n(5):(0,9-0)                  R=  -2.8  p = 0.921     normal           
  mod3n(5):(1,9-0)                  R=  +0.7  p = 0.369     normal           
  mod3n(5):(2,9-0)                  R=  -0.9  p = 0.675     normal           
  mod3n(5):(3,9-0)                  R=  +0.6  p = 0.388     normal           
  mod3n(5):(4,9-0)                  R=  +0.5  p = 0.408     normal           
  mod3n(5):(5,9-0)                  R=  +2.0  p = 0.166     normal           
  mod3n(5):(6,9-0)                  R=  -1.2  p = 0.721     normal           
  mod3n(5):(7,9-0)                  R=  -1.2  p = 0.727     normal           
  mod3n(5):(8,9-1)                  R=  -1.0  p = 0.696     normal           
  mod3n(5):(9,9-1)                  R=  +1.1  p = 0.292     normal           
  mod3n(5):(10,9-2)                 R=  -1.9  p = 0.829     normal           
  mod3n(5):(11,9-2)                 R=  +1.8  p = 0.187     normal           
  mod3n(5):(12,9-3)                 R=  -0.3  p = 0.556     normal           
  mod3n(5):(13,9-3)                 R=  +1.3  p = 0.255     normal           
  mod3n(5):(14,9-4)                 R=  +0.1  p = 0.468     normal           
  mod3n(5):(15,9-4)                 R=  +1.5  p = 0.216     normal           
  TMFn(2+0):wl                      R=  -0.2  p~= 0.5       normal           
  TMFn(2+1):wl                      R=  +0.2  p~= 0.5       normal           
  TMFn(2+2):wl                      R=  -1.3  p~= 0.7       normal           
  TMFn(2+3):wl                      R=  +0.7  p~= 0.4       normal           
  TMFn(2+4):wl                      R=  +1.8  p~= 0.3       normal           
  TMFn(2+5):wl                      R=  +0.8  p~= 0.4       normal           
  TMFn(2+6):wl                      R=  +1.9  p~= 0.3       normal           
  TMFn(2+7):wl                      R=  +2.2  p~= 0.2       normal           
  TMFn(2+8):wl                      R=  -0.9  p~= 0.6       normal           
  TMFn(2+9):wl                      R=  +1.2  p~= 0.3       normal           
  TMFn(2+10):wl                     R=  +1.5  p~= 0.3       normal           
  TMFn(2+11):wl                     R=  +5.2  p~= 0.08      normal           
  [Low1/8]BCFN(2+0,13-0,T)          R=  +1.8  p = 0.232     normal           
  [Low1/8]BCFN(2+1,13-0,T)          R=  +2.0  p = 0.205     normal           
  [Low1/8]BCFN(2+2,13-0,T)          R=  -0.0  p = 0.500     normal           
  [Low1/8]BCFN(2+3,13-0,T)          R=  -1.0  p = 0.657     normal           
  [Low1/8]BCFN(2+4,13-0,T)          R=  +0.7  p = 0.390     normal           
  [Low1/8]BCFN(2+5,13-1,T)          R=  +1.3  p = 0.299     normal           
  [Low1/8]BCFN(2+6,13-1,T)          R=  +1.6  p = 0.248     normal           
  [Low1/8]BCFN(2+7,13-2,T)          R=  +2.7  p = 0.134     normal           
  [Low1/8]BCFN(2+8,13-3,T)          R=  +0.7  p = 0.377     normal           
  [Low1/8]BCFN(2+9,13-3,T)          R=  -3.1  p = 0.903     normal           
  [Low1/8]BCFN(2+10,13-4,T)         R=  -2.5  p = 0.854     normal           
  [Low1/8]BCFN(2+11,13-5,T)         R=  -3.1  p = 0.910     normal           
  [Low1/8]BCFN(2+12,13-5,T)         R=  +3.3  p = 0.095     normal           
  [Low1/8]BCFN(2+13,13-6,T)         R=  -2.8  p = 0.892     normal           
  [Low1/8]BCFN(2+14,13-6,T)         R=  +1.2  p = 0.281     normal           
  [Low1/8]BCFN(2+15,13-7,T)         R=  +1.2  p = 0.277     normal           
  [Low1/8]BCFN(2+16,13-8,T)         R=  +0.8  p = 0.305     normal           
  [Low1/8]BCFN(2+17,13-8,T)         R=  +0.8  p = 0.311     normal           
  [Low1/8]BCFN(2+18,13-9,T)         R=  -2.4  p = 0.908     normal           
  [Low1/8]BCFN(2+19,13-9,T)         R=  +0.4  p = 0.339     normal           
  [Low1/8]DC6-9x1Bytes-1            R=  +0.8  p = 0.401     normal           
  [Low1/8]Gap-16:A                  R=  -0.7  p = 0.746     normal           
  [Low1/8]Gap-16:B                  R=  -0.2  p = 0.556     normal           
  [Low1/8]FPF-14+6/16:(0,14-0)      R=  +1.1  p = 0.224     normal           
  [Low1/8]FPF-14+6/16:(1,14-0)      R=  -1.9  p = 0.911     normal           
  [Low1/8]FPF-14+6/16:(2,14-0)      R=  -1.7  p = 0.889     normal           
  [Low1/8]FPF-14+6/16:(3,14-0)      R=  +0.4  p = 0.398     normal           
  [Low1/8]FPF-14+6/16:(4,14-0)      R=  +2.1  p = 0.069     normal           
  [Low1/8]FPF-14+6/16:(5,14-0)      R=  -0.4  p = 0.601     normal           
  [Low1/8]FPF-14+6/16:(6,14-0)      R=  +2.2  p = 0.062     normal           
  [Low1/8]FPF-14+6/16:(7,14-0)      R=  +0.5  p = 0.359     normal           
  [Low1/8]FPF-14+6/16:(8,14-0)      R=  -0.5  p = 0.631     normal           
  [Low1/8]FPF-14+6/16:(9,14-1)      R=  -2.3  p = 0.949     normal           
  [Low1/8]FPF-14+6/16:(10,14-2)     R=  -0.6  p = 0.675     normal           
  [Low1/8]FPF-14+6/16:(11,14-2)     R=  -1.4  p = 0.843     normal           
  [Low1/8]FPF-14+6/16:(12,14-3)     R=  -0.5  p = 0.630     normal           
  [Low1/8]FPF-14+6/16:(13,14-4)     R=  -0.6  p = 0.655     normal           
  [Low1/8]FPF-14+6/16:(14,14-5)     R=  +0.2  p = 0.443     normal           
  [Low1/8]FPF-14+6/16:(15,14-5)     R=  -0.7  p = 0.698     normal           
  [Low1/8]FPF-14+6/16:(16,14-6)     R=  -0.1  p = 0.530     normal           
  [Low1/8]FPF-14+6/16:(17,14-7)     R=  -1.5  p = 0.854     normal           
  [Low1/8]FPF-14+6/16:(18,14-8)     R=  +1.0  p = 0.223     normal           
  [Low1/8]FPF-14+6/16:(19,14-8)     R=  -1.1  p = 0.778     normal           
  [Low1/8]FPF-14+6/16:(20,14-9)     R=  +0.9  p = 0.252     normal           
  [Low1/8]FPF-14+6/16:(21,14-10)    R=  -0.2  p = 0.522     normal           
  [Low1/8]FPF-14+6/16:(22,14-11)    R=  +0.4  p = 0.340     normal           
  [Low1/8]FPF-14+6/16:(23,14-11)    R=  +1.7  p = 0.116     normal           
  [Low1/8]FPF-14+6/16:all           R=  -0.4  p = 0.615     normal           
  [Low1/8]FPF-14+6/16:cross         R=  -0.3  p = 0.565     normal           
  [Low1/8]BRank(12):128(8)          R=  +1.5  p~= 0.080     normal           
  [Low1/8]BRank(12):256(8)          R=  -0.7  p~= 0.670     normal           
  [Low1/8]BRank(12):384(2)          R=  +0.8  p~= 0.293     normal           
  [Low1/8]BRank(12):512(4)          R=  -0.8  p~= 0.670     normal           
  [Low1/8]BRank(12):768(1)          R=  +0.4  p~= 0.366     normal           
  [Low1/8]BRank(12):1K(4)           R=  -1.4  p~= 0.890     normal           
  [Low1/8]BRank(12):1536(1)         R=  -0.7  p~= 0.689     normal           
  [Low1/8]BRank(12):2K(2)           R=  -0.2  p~= 0.554     normal           

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2021-06-15
### Fixed
- republish new version with correct compiler target for java 8

## [1.0.0] - 2021-06-14 Changed groupId and reset version

### Changed
- change groupId to dev.brachtendorf
- distribute via maven central

## [1.0.1] - 2019-01-12

### Changed
- swapped oraclejdk for openjdk to support travis fix
- The previous version of pcgRandom did not utilize the setSeed method found in Random, due to the 
fact that the rng is depended on 2 variables. The stream number and the seed. Override the setSeed 
method to while prominently nothing that this won't produce
the same instance exactly the same as using the seed in the constructor.
- Generalized Junit test cases to extends a base class
- Upgraded maven source and javadoc plugin to new version

### Added
 - Test casese to cover the new setSeed method
 - setSeed and setSeed persistent
 - codacay support
 
### Fixed
- PCGRsFast now correctly throws a IllegalArgumentException if nextBoolean(probaility) and nextLong(n) is out of bounds
- various code smells

## [1.0.0] - 2018-10-05 Initial release


src/main/java/com/github/kilianB/pcg/RandomBase64.java
package com.github.kilianB.pcg;

import java.lang.reflect.InvocationTargetException;
import java.util.Random;

/**
 * Base class for 64 bit state pcg random number generators with 32 bit output.
 * 
 * The PCG family uses a linear congruential generator as the state-transition
 * function—the “CG” of PCG stands for “congruential generator”. Linear
 * congruential generators are known to be statistically weak.
 * <p>
 * 
 * PCG uses a new technique called permutation functions on tuples to produce
 * output that is much more random than the RNG's internal state. The output
 * function is defined by the extending classes.
 * <p>
 * 
 * A paper highlighting the individual properties can be found here. <a href=
 * "http://www.pcg-random.org/paper.html">http://www.pcg-random.org/paper.html</a>.
 * This class is an adaption to the original c
 * <a href="https://github.com/imneme/pcg-c">source code</a> provided by M.E.
 * O'Neill.
 * <p>
 *
 * <b>Contract:</b> every extending class <b>must</b> implement a copy
 * constructor with a signature of(long,long,boolean). As it does not perform
 * proper initialization of the seed this method should not be exposed.
 * 
 * @author Kilian
 *
 * @see <a href="http://www.pcg-random.org/">www.pcg-random.org</a>
 */
public abstract class RandomBase64 extends Random implements Pcg {

	private static final long serialVersionUID = -4396858403047759432L;

	/**
	 * Linear congruential constant. Same as MMIX by Donald Knuth and Newlib, Musl
	 */
	protected static final long MULT_64 = 6364136223846793005L;

	/**
	 * Seeds the generator with 2 longs generated by xorshift*. The values choosen
	 * are very likely not used in any other invocation of this constructor.
	 */
	public RandomBase64() {
		this(getRandomSeed(), getRandomSeed());
	}

	/**
	 * Create a random number generator with the given seed and stream number. The
	 * seed defines the current state in which the rng is in and corresponds to
	 * seeds usually found in other RNG instances. RNGs with different seeds are
	 * able to catch up after they exhaust their period and produce the same
	 * numbers. (2^63).
	 * <p>
	 * 
	 * Different stream numbers alter the increment of the rng and ensure distinct
	 * state sequences
	 * <p>
	 * 
	 * Only generators with the same seed AND stream numbers will produce identical
	 * values
	 * <p>
	 * 
	 * @param seed         used to compute the starting state of the RNG
	 * @param streamNumber used to compute the increment for the lcg.
	 */
	public RandomBase64(long seed, long streamNumber) {
		setSeed(seed, streamNumber);
	}

	/**
	 * Copy constructor. <b>Has</b> to be implemented in all inheriting instances.
	 * This will be invoked through reflection! when calling {@link #split()} or
	 * {@link #splitDistinct()} If no special behavior is desired simply pass though
	 * the values.
	 * 
	 * This constructor should usually not be called manually as the seed and
	 * increment will just be set without performing any randomization.
	 * 
	 * @param initialState of the lcg. The value will be set and not altered.
	 * @param increment    used in the lcg. has to be odd
	 * @param dummy        unused. Resolve signature disambiguate
	 */
	@Deprecated
	protected RandomBase64(long initialState, long increment, boolean dummy) {

		if (increment == 0) {
			throw new IllegalArgumentException("The increment can't be 0");
		}

		if (increment % 2 == 0) {
			throw new IllegalArgumentException("Increment has to be odd");
		}

		// Use getters and setters to let fast implementation overwrite this behavior
		// while still maintaining inheritance.
		setState(initialState);
		setInc(increment);
	}

	/**
	 * Sets the seed of this random number generator using a single long seed. The
	 * general contract of setSeed is that it alters the state of this random number
	 * generator object so as to be in exactly the same state as if it had just been
	 * created with the argument seed as a seed.
	 * <p>
	 * Be aware that seeding this random number generation requires 2 arguments. A
	 * seed and a stream number.
	 * <p>
	 * Calling this method is equivalent to {@link RandomBase64(seed,seed)};
	 * Confusion may arise when constructing a rng instance with a given seed and a
	 * different stream number and expecting the rngs to be in the same state as
	 * right after the constructor call.
	 * <p>
	 * In this case please refer to the method {@link #setSeed(long, long)};
	 * 
	 * @deprecated This method behaves differently than you would expect from the
	 *             random base class.
	 * @since 1.0.1
	 */
	@Override
	public void setSeed(long seed) {
		setSeed(seed, seed);
	}

	/**
	 * Sets the seed of this random number generator using . The general contract of
	 * setSeed is that it alters the state of this random number generator object so
	 * as to be in exactly the same state as if it had just been created with the
	 * argument seed as a seed.
	 * 
	 * Only generators with the same seed AND stream numbers will produce identical
	 * values
	 * <p>
	 * 
	 * @param seed         used to compute the starting state of the RNG
	 * @param streamNumber used to compute the increment for the lcg.
	 */
	public void setSeed(long seed, long streamNumber) {
		setState(0);
		/*
		 * We need to ensure that subclasses can override and add synchronization as
		 * they please. use getters and setters.
		 * 
		 * Although there are rules for the choice of constants [17], if we pick a
		 * power-of-two modulus and a good multiplicative constant, the only constraint
		 * on c for a full period generator is that c is odd and > 0
		 * 
		 * Chapter 4.2.1 (http://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf)
		 */
		setInc((streamNumber << 1) | 1); // 2* + 1
		stepRight();
		setState(getState() + seed);
		stepRight();
	}

	/**
	 * Update the state of the lcg and move a step forward. The old state should be
	 * used to extract bits used to construct a number.
	 * <p>
	 * 
	 * When implementations use the newly generate state to calculate random numbers
	 * {@link #isFast()} has to return true.
	 * 
	 * @return the old value of the state variable before updating.
	 */
	protected abstract long stepRight();

	@Override
	@SuppressWarnings("unchecked")
	public <T> T split() throws ReflectiveOperationException {
		try {
			return (T) getClass().getDeclaredConstructor(long.class, long.class, boolean.class).newInstance(getState(),
					getInc(), true);
		} catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
				| NoSuchMethodException | SecurityException e) {
			e.printStackTrace();
			throw new ReflectiveOperationException("Failed to instantiate clone constructor");
		}
	}

	@Override
	@SuppressWarnings("unchecked")
	public <T> T splitDistinct() throws ReflectiveOperationException {
		try {
			long curInc;
			long curState;

			// No reason to CAS here. we don't swap the inc around all the time
			do {
				// Has to be odd
				curInc = ((nextLong(Math.abs(getInc())) ^ (~System.nanoTime())) * 2) + 1;
			} while (curInc == getInc());

			// State swaps by each call to nextLong
			do {
				curState = (nextLong(Math.abs(getState())) ^ (~System.nanoTime()));
			} while (curState == getState());

			return (T) getClass().getDeclaredConstructor(long.class, long.class, boolean.class).newInstance(curState,
					curInc, true);
		} catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
				| NoSuchMethodException | SecurityException e) {
			e.printStackTrace();
			e.getCause().printStackTrace();
			throw new ReflectiveOperationException("Failed to instantiate clone constructor");
		}
	}

	@Override
	public int next(int n) {
		int nInt = nextInt();
		int shifted = (nInt >>> (32 - n));
		return shifted;
	}

	/**
	 * Construct a 32bit int from the given 64bit state using a permutation
	 * function. The produced int will be used to construct all other datatypes
	 * returned by this RNG.
	 * 
	 * @param state random int as produced by the internal lcg
	 * @return a random int with randomly set bits
	 * 
	 */
	protected abstract int getInt(long state);

	// Non standard random functions. add missing convinience methods

	@Override
	public boolean nextBoolean(double probability) {
		if (probability < 0.0 || probability > 1.0)
			throw new IllegalArgumentException("probability must be between 0.0 and 1.0 inclusive.");
		if (probability == 0.0)
			return false; // fix half-open issues
		else if (probability == 1.0)
			return true; // fix half-open issues
		return nextDouble() < probability;
	}

	@Override
	public byte nextByte() {
		return (byte) (next(8));
	}

	@Override
	public void nextBytes(byte[] bytes) {
		// According to merseene twister
		/*
		 * A bug fix for all versions of the JDK. The JDK appears to use all four bytes
		 * in an integer as independent byte values! Totally wrong. I've submitted a bug
		 * report.
		 */
		for (int x = 0; x < bytes.length; x++)
			bytes[x] = (byte) next(8);
	}

	@Override
	public char nextChar() {
		return (char) (next(16));
	}

	@Override
	public short nextShort() {
		return (short) (next(16));
	}

	@Override
	public int nextInt() {
		return getInt(stepRight());
	}

	/**
	 * {@inheritDoc}
	 * <p>
	 * The long value is composed of two integers and a 64 bit state therefore,
	 * returning all possible long values with equal probability.
	 * 
	 */
	public long nextLong() {
		// Don't override just update javadocs
		return super.nextLong();
	}

	/**
	 * {@inheritDoc}
	 * <p>
	 * The long value is composed of two integers and a 64 bit state therefore,
	 * returning all possible long values with equal probability.
	 */
	@Override
	public long nextLong(long n) {
		if (n <= 0)
			throw new IllegalArgumentException("n must be positive, got: " + n);
		long bits;
		long val;
		do {
			bits = (nextLong() >>> 1);
			val = bits % n;
		} while (bits - val + (n - 1) < 0);
		return val;
	}

	@Override
	public float nextFloat(boolean includeZero, boolean includeOne) {
		float d = 0.0f;
		do {
			d = nextFloat(); // grab a value, initially from half-open [0.0f, 1.0f)
			if (includeOne && nextBoolean())
				d += 1.0f; // if includeOne, with 1/2 probability, push to [1.0f, 2.0f)
		} while ((d > 1.0f) || // everything above 1.0f is always invalid
				(!includeZero && d == 0.0f)); // if we're not including zero, 0.0f is invalid
		return d;
	}

	@Override
	public double nextDouble(boolean includeZero, boolean includeOne) {
		double d = 0.0;
		do {
			d = nextDouble(); // grab a value, initially from half-open [0.0, 1.0)
			if (includeOne && nextBoolean())
				d += 1.0; // if includeOne, with 1/2 probability, push to [1.0, 2.0)
		} while ((d > 1.0) || // everything above 1.0 is always invalid
				(!includeZero && d == 0.0)); // if we're not including zero, 0.0 is invalid
		return d;
	}

	/*
	 * Protected static isn't really "clean" but roll with it. No reason to make it
	 * public and define it as a default method in the interface as it is defined
	 * only for the 64 bit state. We want to have control over every instance
	 * therefor take the static path. The fast instance needs access to it. Else we
	 * could also provide package visbility or use reflection
	 */
	/**
	 * Return a random 64 bit seed ensuring uniqueness by using a xorshift64* shift
	 * algorithm. This implementation is thread safe.
	 * 
	 * @return a unique seed
	 */
	protected static long getRandomSeed() {
		// xorshift64*
		for (;;) {
			long current = UNIQUE_SEED.get();
			long next = current;
			next ^= next >> 12;
			next ^= next << 25; // b
			next ^= next >> 27; // c
			next *= 0x2545F4914F6CDD1DL;
			if (UNIQUE_SEED.compareAndSet(current, next))
				return next;
		}
	}

	// Below add some support for fast instances which almost always do not rely on
	// the defined methods in this class but still extend it to allow for
	// polymorphism.

	@Override
	public long getMult() {
		return MULT_64;
	}

	/**
	 * Set the internal state of the pcg. This method is used during the seeding
	 * process of this class and therefore, it is most likely is never correct to
	 * alter the variable passed to this function.
	 * <p>
	 * Allowed operations are synchronization on those methods-
	 * <p>
	 * 
	 * @param newState of the pcg
	 */
	protected abstract void setState(long newState);

	/**
	 * Set the increment of the pcg. This method is used during the seeding process
	 * of this class and therefore, it is most likely is never correct to alter the
	 * variable passed to this function.
	 * <p>
	 * "Although there are rules for the choice of constants [17], if we pick a
	 * power-of-two modulus and a good multiplicative constant, the only constraint
	 * on c for a full period generator is that c is odd and {@literal>} 0"
	 * <p>
	 * Chapter 4.2.1 (http://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf)
	 * <p>
	 * Allowed operations are synchronization on those methods-
	 * 
	 * @param newInc of the pcg
	 */
	protected abstract void setInc(long newInc);
}
